src/aspire/ctf/ctf_estimator.py

```diff
@@ -0,0 +1,757 @@
+"""
+Created on Sep 10, 2019
+@author: Ayelet Heimowitz, Amit Moscovich
+
+Integrated into ASPIRE by Garrett Wright Feb 2021.
+"""
+
+import logging
+import os
+
+import numpy as np
+from numpy import linalg as npla
+from scipy.optimize import linprog
+
+from aspire.image import Image
+from aspire.numeric import fft
+from aspire.utils import complex_type, eigs
+
+logger = logging.getLogger(__name__)
+
+
+def abs2(x):
+    """
+    Compute complex modulus squared.
+    """
+
+    return x.real ** 2 + x.imag ** 2
+
+
+class CtfEstimator:
+    """
+    CtfEstimator Class ...
+    """
+
+    def __init__(
+        self,
+        pixel_size,
+        cs,
+        amplitude_contrast,
+        voltage,
+        psd_size,
+        num_tapers,
+        dtype=np.float32,
+    ):
+        """
+        Instantiate a CtfEstimator instance.
+
+        :param pixel_size: Size of the pixel in Angstrom.
+        :param cs: Spherical aberration in mm.
+        :param amplitude_contrast: Amplitude contrast.
+        :param voltage: Voltage of electron microscope.
+        :param psd_size: Block size (in pixels) for PSD estimation.
+        :param num_tapers: Number of tapers to apply in PSD estimation.
+        :returns: CtfEstimator instance.
+        """
+
+        self.pixel_size = pixel_size
+        self.cs = cs
+        self.amplitude_contrast = amplitude_contrast
+        self.voltage = voltage
+        self.psd_size = psd_size
+        self.num_tapers = num_tapers
+        self.lmbd = 1.22639 / np.sqrt(voltage * 1000 + 0.97845 * np.square(voltage))
+        self.dtype = np.dtype(dtype)
+
+        center = psd_size // 2
+        [X, Y] = np.meshgrid(
+            np.arange(0 - center, psd_size - center, dtype=self.dtype) / psd_size,
+            np.arange(0 - center, psd_size - center, dtype=self.dtype) / psd_size,
+        )
+
+        rb = np.sqrt(np.square(X) + np.square(Y))
+
+        self.r_ctf = rb * (10 / pixel_size)
```

Comment:
I agree with adding comment, but I don't follow the units. Pixel_size was doc'd to be angstrom, so rhs:

10 / p_angstrom = 10 / (p_angstrom * 1nm/(10angstrom) )= 10 * 10 / p _nm

Its not clear to me that rb unit is nm^2/100.  Want to make sure I add comment correctly...

Reply:
Sorry it should be inverse nanometers. If p is in Å/pixel, then p/10 is in nm/pixel and consequently 10/p is in pixel/nm. Since rb is in inverse pixels, r_ctf should be in inverse nm.

Reply:
cool, thanks, added a note.